src/main/scala/org/ergoplatform/nodeView/state/ErgoState.scala

```diff
@@ -152,20 +151,38 @@ object ErgoState extends ScorexLogging {
     *         if box was first spend and created after that - it is in both toInsert and toRemove,
     *         and an error will be thrown further during tree modification
     */
-  def boxChanges(txs: Seq[ErgoTransaction]): (Seq[ADKey], Seq[ErgoBox]) = {
-    val toInsert: mutable.HashMap[ModifierId, ErgoBox] = mutable.HashMap.empty
-    val toRemove: mutable.ArrayBuffer[(ModifierId, ADKey)] = mutable.ArrayBuffer()
-    txs.foreach { tx =>
+  def boxChanges(txs: Seq[ErgoTransaction]): Try[(Vector[Remove], Vector[Insert])] = Try {
+    val toInsert: mutable.TreeMap[ModifierId, Insert] = mutable.TreeMap.empty
+    val toRemove: mutable.TreeMap[ModifierId, Remove] = mutable.TreeMap.empty
+
+    cfor(0)(_ < txs.length, _ + 1) { i =>
+      val tx = txs(i)
       tx.inputs.foreach { i =>
-        val wrapped = bytesToId(i.boxId)
-        toInsert.remove(wrapped) match {
-          case None => toRemove.append((wrapped, i.boxId))
+        val wrappedBoxId = bytesToId(i.boxId)
+        toInsert.remove(wrappedBoxId) match {
+          case None =>
+            if (toRemove.put(wrappedBoxId, Remove(i.boxId)).nonEmpty) {
+              throw new IllegalArgumentException(s"Tx : ${tx.id} is double-spending input id : $wrappedBoxId")
```

Comment:
Is it part of consensus? This new implementation may throw when the old don't.

Reply:
The code is consensus-critical, the old version also throws exception, but during tree modification stage